Implemented "Emergency Exit" and "Wimp Out"

Korwai opened the pull request:
Added Wimpod's and Golisopod's ability to forcibly retreat upon reaching 50% HP or less.

- https://bulbapedia.bulbagarden.net/wiki/Emergency_Exit_(Ability)

I've introduced PostDamageAbAttrs since when I initially tried with PostDefend, it was activating before the HP bar had the chance to update which looked off. PostDamageAbAttr should kick in after damageAndUpdate() which should make more sense for this ability visually.

I've tested that it works for the player's & enemy's Golisopod in singles and doubles (wild and trainer). ~~Berries also activate before the HP condition check.~~

![emergency exit](https://github.com/pagefaultgames/pokerogue/assets/24775950/6677450b-5792-428b-83bb-56138713d050)

bennybroseph commented:
Would like to check this out if the conflicts can be resolved. Seems like reverse U-Turn which is not behaving nicely like in your video.

bennybroseph reviewed (changes requested):
Looking to get JSDoc documentation as shown here: [Code Comment FAQ](https://discord.com/channels/1125469663833370665/1239957703820836975)

Korwai commented on src/data/ability.ts:
```diff
@@ -443,12 +443,62 @@ export class NonSuperEffectiveImmunityAbAttr extends TypeImmunityAbAttr {
+        // Wild pokemon
+        pokemon.hideInfo().then(() => pokemon.destroy());
+	  	  pokemon.scene.queueMessage(getPokemonMessage(pokemon, ' fled!'), null, true, 500);
+        pokemon.scene.pushPhase(new BattleEndPhase(pokemon.scene));
```
Overlooked a few conditions here. I can't blindly push the end of battle phase if it's a double battle. Gotta check both pokemon are either not active or fainted, I think.

I might also have to look into the condition if pokemon 1 faints but pokemon 2 (Golisopod) flees. It should go to the reward phase.

Korwai commented on src/phases.ts:
```diff
@@ -2543,7 +2543,7 @@ export class MoveEffectPhase extends PokemonPhase {
                           target.addTag(BattlerTagType.FLINCHED, undefined, this.move.moveId, user.id);
                       }
-                      Utils.executeIf(!isProtected && !chargeEffect, () => applyFilteredMoveAttrs((attr: MoveAttr) => attr instanceof MoveEffectAttr && (attr as MoveEffectAttr).trigger === MoveEffectTrigger.HIT && (!attr.firstHitOnly || firstHit),
+                      Utils.executeIf(!isProtected && !chargeEffect && target.active, () => applyFilteredMoveAttrs((attr: MoveAttr) => attr instanceof MoveEffectAttr && (attr as MoveEffectAttr).trigger === MoveEffectTrigger.HIT && (!attr.firstHitOnly || firstHit),
```
Worried if this would cause trouble anywhere else but ultimately, if a pokemon flees, it's no longer active and we shouldn't check postDefend/postAttackAbAttrs after.

I was debating adding another status effect called FLED so that this behavior wouldn't affect anything else but I wasn't sure if it belonged as a status effect. Any thoughts on if this should be changed?

Korwai commented on src/data/ability.ts:
```diff
@@ -443,12 +443,92 @@ export class NonSuperEffectiveImmunityAbAttr extends TypeImmunityAbAttr {
+   * @param initialPokemonHpRatio The initial HP ratio of the pokemon that's being forced to switch/flee.
+   * @param passive 
+   * @returns true if the function succeeds.
+   */
```
Honestly, I'm not sure what 'passive' is since I copied it from the other AbAttr classes (that also don't make use of it).

Korwai commented:
Added JSDoc comments to most of the big classes/functions that I've added. I've also adjusted the logic for double battles for enemies. Happy to change anything if needed.

flx-sta commented:
I've tried it with 2 Pokemon (for both abilities) but neither of them wanted to use their ability ❌ 

<img width="986" alt="image" src="https://github.com/pagefaultgames/pokerogue/assets/50131232/1500ac6f-98e2-4bb3-812c-235aea483cab">
<img width="1009" alt="image" src="https://github.com/pagefaultgames/pokerogue/assets/50131232/98e350d7-a2a4-4c03-aaa3-dd35a34f7274">

<img width="1009" alt="image" src="https://github.com/pagefaultgames/pokerogue/assets/50131232/63c1377e-dbec-4c3f-b128-ac35422a8434">
<img width="999" alt="image" src="https://github.com/pagefaultgames/pokerogue/assets/50131232/b81e8df6-9935-418c-b7aa-45139fc0f3a9">

flx-sta reviewed (changes requested):
### I've tried it with 2 Pokemon (for both abilities), against Wild Pkmn and trainer but neither of them switch out ❌ 

<img width="350" alt="image" src="https://github.com/pagefaultgames/pokerogue/assets/50131232/1500ac6f-98e2-4bb3-812c-235aea483cab"><img width="350" alt="image" src="https://github.com/pagefaultgames/pokerogue/assets/50131232/98e350d7-a2a4-4c03-aaa3-dd35a34f7274">

<img width="350" alt="image" src="https://github.com/pagefaultgames/pokerogue/assets/50131232/63c1377e-dbec-4c3f-b128-ac35422a8434"><img width="350" alt="image" src="https://github.com/pagefaultgames/pokerogue/assets/50131232/b81e8df6-9935-418c-b7aa-45139fc0f3a9">

Korwai commented:
Gah, I removed the line of code that was supposed to trigger the ability. That's my bad for not testing thoroughly before pushing. I'll update it and test it again.

Korwai commented on src/data/ability.ts:
```diff
@@ -460,12 +460,102 @@ export class NonSuperEffectiveImmunityAbAttr extends TypeImmunityAbAttr {
+        }
+      } else if (this.condition(pokemon, initialPokemonHpRatio) && !pokemon.hasTrainer()) {
+        // Wild pokemon
+        pokemon.scene.time.delayedCall(200, () => {
```
I added this in because it was difficult to tell the final HP of the pokemon visually after a hit since the ability triggers immediately after. This make it easier to read at lower game speeds (1x, 2x) but anything higher, it looks the same visually.

Not sure if this is the right way to go about doing this because I believe this is triggering this function 200ms later which can lead to unexpected behavior.

flx-sta replied:
Hi don’t k ow what to think about this. I get the idea but hard-coded delays are … ouf…

Korwai replied:
Yeah... thinking about this more... this is a terrible way to do it.

I now know of a different way though. Triggering the ability during the TurnEndPhase would allow MoveEndPhase and BerryPhase messages to play before and should resolve this issue.

Korwai commented:
Here's some sample videos of my testing.

Enemy trainers properly switch with parties of 3+ non-fainted mons in double battles.
https://github.com/pagefaultgames/pokerogue/assets/24775950/a03335ab-1866-414f-a6fc-4de294f79610

As long as one wild mon dies in a double battle, it will trigger a reward phase.
https://github.com/pagefaultgames/pokerogue/assets/24775950/7a841076-c1dc-4981-bcad-ad042c0d847a

Enemy trainers properly switch with parties of 2+ non-fainted mons in single battles. Also, status effects don't trigger the forced switch.
https://github.com/pagefaultgames/pokerogue/assets/24775950/c5e0c2e6-259c-4f44-81b7-eb2dfd28beb6

I'm aware right now that berries do not trigger before the check but they still get consumed.

OrangeRed commented on src/data/ability.ts:
```diff
@@ -3434,6 +3524,11 @@ export function applyPostDefendAbAttrs(attrType: { new(...args: any[]): PostDefe
   return applyAbAttrsInternal<PostDefendAbAttr>(attrType, pokemon, (attr, passive) => attr.applyPostDefend(pokemon, passive, attacker, move, hitResult, args), args);
 }
 
+export function applyPostDamageAbAttrs(attrType: { new(...args: any[]): PostDamageAbAttr },
+  pokemon: Pokemon, initialPokemonHpRatio: integer, ...args: any[]): Promise<void> {
+  return applyAbAttrsInternal<PostDamageAbAttr>(attrType, pokemon, (attr, passive) => attr.applyPostDamage(pokemon, initialPokemonHpRatio, passive), args);
+}
+
```
Is there any reason why this has to be `PostDamageAbAttr` and can't be a `PostMoveUsedAbAttr`?

Korwai replied:
From what I understand, the MovePhase triggers before the DamagePhase so I don't think it'd be able to check for different HP values from pre/post hits.

OrangeRed replied:
PostTurnAbAttr should trigger after berries no?

Korwai replied:
Mistyped but yes. I think that's a better path.

OrangeRed commented on src/data/ability.ts:
```diff
@@ -460,12 +460,102 @@ export class NonSuperEffectiveImmunityAbAttr extends TypeImmunityAbAttr {
+   * @returns true if the function succeeds.
+   */
+  applyPostDamage(pokemon: Pokemon, initialPokemonHpRatio: number, passive: boolean): Promise<boolean> {
+    return new Promise<boolean>(resolve => {
```
Does this need to be a promise? Can't this be rewritten to simply return true / false instead?

OrangeRed commented on src/data/ability.ts:
```diff
@@ -460,12 +460,102 @@ export class NonSuperEffectiveImmunityAbAttr extends TypeImmunityAbAttr {
+   */
+  applyPostDamage(pokemon: Pokemon, initialPokemonHpRatio: number, passive: boolean): Promise<boolean> {
+    return new Promise<boolean>(resolve => {
+      pokemon.scene.unshiftPhase(new BerryPhase(pokemon.scene));
```
Why are you triggering berry phase here? Doesn't berry phase trigger on its own at the end of the turn?

Korwai replied:
Forgot to remove, was part of my testing with berries not triggering before the check.

flx-sta replied:
@Korwai advice for next time:
Put a `// TODO: remove` behind it, on the top of it

Korwai commented:
More testing and cleaning, hopefully it's a bit less buggy now.

Multi-hits only trigger Emergency Exit/Wimp Out on the last hit. Berries will also proc before the ability checks HP.
<video src="https://github.com/pagefaultgames/pokerogue/assets/24775950/f6495cfc-d260-4ebe-8a8c-90c624f89c61
" width="300" />

U-Turn will not trigger if it triggers the opponent's Emergency Exit/Wimp Out skill.
<video src="https://github.com/pagefaultgames/pokerogue/assets/24775950/ffaf2225-d113-458b-b58f-afa65d7cab39
" width="300" />

EvasiveAce commented:
Does your Wimp Out/Emergency Exit allow you to choose who you switch in to, or is it random?

flx-sta commented:
> Does your Wimp Out/Emergency Exit allow you to choose who you switch in to, or is it random?

You can see that behavior on his [first comment](https://github.com/pagefaultgames/pokerogue/pull/384#issue-2274090909) in the gif.

EvasiveAce commented:
> > Does your Wimp Out/Emergency Exit allow you to choose who you switch in to, or is it random?
> 
> You can see that behavior on his [first comment](https://github.com/pagefaultgames/pokerogue/pull/384#issue-2274090909) in the gif.

I know, I was just seeing if its changed. If not, add me on Discord because I have code for it to be random. @Korwai 

EvasiveAce

Korwai commented:
> > > Does your Wimp Out/Emergency Exit allow you to choose who you switch in to, or is it random?
> > 
> > 
> > You can see that behavior on his [first comment](https://github.com/pagefaultgames/pokerogue/pull/384#issue-2274090909) in the gif.
> 
> I know, I was just seeing if its changed. If not, add me on Discord because I have code for it to be random. @Korwai
> 
> EvasiveAce

Still the same behaviour as the first gif. I'm pretty sure Emergency Exit/Wimp Out allows you to choose which pokemon you want to swap into if the HP condition is met.

Source: https://youtu.be/3MBXB7hozaA?si=AWXmxAwY2MJpB6oz&t=985 (audio/volume warning)

EvasiveAce commented:
> > > > Does your Wimp Out/Emergency Exit allow you to choose who you switch in to, or is it random?
> > > 
> > > 
> > > You can see that behavior on his [first comment](https://github.com/pagefaultgames/pokerogue/pull/384#issue-2274090909) in the gif.
> > 
> > 
> > I know, I was just seeing if its changed. If not, add me on Discord because I have code for it to be random. @Korwai
> > EvasiveAce
> 
> Still the same behaviour as the first gif. I'm pretty sure Emergency Exit/Wimp Out allows you to choose which pokemon you want to swap into if the HP condition is met.
> 
> Source: https://youtu.be/3MBXB7hozaA?si=AWXmxAwY2MJpB6oz&t=985 (audio/volume warning)

Damn that sucks they changed it. Anyways good PR

Korwai commented:
Well, I messed up that rebase. Closing this PR for a cleaner slate.

flx-sta commented:
Please link the slate
